### PR TITLE
Remove all unsafe IO from Array.Foreign.Mut.Type

### LINKED ISSUE #1283: Remove unsafePerformIO from mutable array module
Fix all the todos from this commit: https://github.com/composewell/streamly/pull/1280/commits/135afd068b6e30aea178699c552b139222904567

### PR DESCRIPTION
**Pull request by adithyaov**
Closes #1283

### CODE REVIEW

**Comment by adithyaov** [comment]
There is a little duplication. Thoughts on how to decrease it?

File: src/Streamly/Internal/Data/Array/Foreign.hs [hunk 1]
```diff
@@ -195,18 +196,41 @@ fromStream (SerialT m) = P.fold A.write m
 -- Elimination
 -------------------------------------------------------------------------------
 
+{-# INLINE_NORMAL producer #-}
```
Comment (rnjtranjan) [inline src/Streamly/Internal/Data/Array/Foreign.hs:199]: It looks like we can re-use " producer :: forall m a. (Monad m, Storable a) => Producer m (Array a) a "
from "src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs"

File: src/Streamly/Internal/Data/Array/Foreign.hs [hunk 2]
```diff
@@ -195,18 +196,41 @@ fromStream (SerialT m) = P.fold A.write m
...
 --
 -- @since 0.8.0
 {-# INLINE_NORMAL read #-}
 read :: forall m a. (Monad m, Storable a) => Unfold m (Array a) a
-read = Unfold.lmap A.unsafeThaw MA.read
-
-{-# INLINE_NORMAL producer #-}
-producer :: forall m a. (Monad m, Storable a) => Producer m (Array a) a
-producer = Producer.translate A.unsafeThaw A.unsafeFreeze MA.producer
+read = Producer.simplify producer
```
Comment (rnjtranjan) [inline src/Streamly/Internal/Data/Array/Foreign.hs:233]: similar code exists in "src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs"

File: src/Streamly/Internal/Data/Array/Foreign.hs (hunk 1 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/Array/Foreign.hs:199]: In `src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs` we no longer use `unsafeInlineIO`. Till I figure out its implication on the performance, I've just duplicated it here.

File: src/Streamly/Internal/Data/Array/Foreign.hs (hunk 2 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/Array/Foreign.hs:233]: In src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs we no longer use unsafeInlineIO. Till I figure out its implication on the performance, I've just duplicated it here.

**Comment by adithyaov** [comment]
The performance seems to be the same.
```
Generating reports for Data.Array.Foreign...
Data.Array.Foreign(cpuTime)
Benchmark                                                                default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------ -------------- --------------------------
All.Data.Array.Foreign/o-1-space.transformationX4.map                            535.81                     +39.20
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.10                       114.43                     +28.61
All.Data.Array.Foreign/o-1-space.elimination./=                                  140.84                     +28.55
All.Data.Array.Foreign/o-1-space.elimination.==                                  149.91                     +15.93
All.Data.Array.Foreign/o-1-space.transformation.map                              172.96                     +14.21
All.Data.Array.Foreign/o-1-space.generation.write . intFromTo                    187.92                      +4.98
All.Data.Array.Foreign/o-1-space.transformation.scanl1'                          252.96                      +4.17
All.Data.Array.Foreign/o-1-space.generation.writeN . fromList                   1018.20                      +3.30
All.Data.Array.Foreign/o-1-space.generation.read                              209545.00                      +3.11
All.Data.Array.Foreign/o-1-space.elimination.<                                   256.01                      +2.84
All.Data.Array.Foreign/o-1-space.transformation.scanl'                           267.91                      +2.27
All.Data.Array.Foreign/o-1-space.elimination.toStreamRev                         198.72                      +2.22
All.Data.Array.Foreign/o-1-space.generation.writeN . unfoldr                     105.82                      +1.67
All.Data.Array.Foreign/o-1-space.elimination.foldl'                              209.37                      +1.24
All.Data.Array.Foreign/o-1-space.elimination.length . IsList.toList              267.47                      +0.95
All.Data.Array.Foreign/o-1-space.elimination.min                                 255.84                      +0.60
All.Data.Array.Foreign/o-1-space.generation.show                               39453.30                      +0.54
All.Data.Array.Foreign/o-1-space.elimination.id                                  153.12                      +0.21
All.Data.Array.Foreign/o-1-space.generation.writeN . IsString.fromString        1173.52                      +0.01
All.Data.Array.Foreign/o-n-space.elimination.writeN                              186.87                      -0.32
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.1                        115.46                      -0.33
All.Data.Array.Foreign/o-n-space.elimination.writeLastN.Max                      303.89                      -1.58
All.Data.Array.Foreign/o-1-space.generation.fromList . intFromTo                1248.27                      -1.84
All.Data.Array.Foreign/o-1-space.transformationX4.scanl'                         772.32                      -3.14
All.Data.Array.Foreign/o-1-space.generation.writeN . intFromTo                    80.60                      -7.51
All.Data.Array.Foreign/o-1-space.elimination.read                                199.76                      -8.87
All.Data.Array.Foreign/o-1-space.transformationX4.scanl1'                        586.36                     -11.12
All.Data.Array.Foreign/o-1-space.generation.writeN . IsList.fromList            1500.30                     -11.30

Data.Array.Foreign(Allocated)
Benchmark                                                                default(0)(Bytes) default(1) - default(0)(%)
------------------------------------------------------------------------ ----------------- --------------------------
All.Data.Array.Foreign/o-1-space.elimination.toStreamRev                         838237.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.1                             0.00                        NaN
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.10                            0.00                        NaN
All.Data.Array.Foreign/o-1-space.transformationX4.map                           4000968.00                      +7.50
All.Data.Array.Foreign/o-1-space.transformationX4.scanl'                        4140230.00                      +3.86
All.Data.Array.Foreign/o-1-space.elimination./=                                  819398.00                      +2.29
All.Data.Array.Foreign/o-1-space.elimination.==                                  819398.00                      +2.29
All.Data.Array.Foreign/o-1-space.generation.read                              542616570.00                      +0.00
All.Data.Array.Foreign/o-1-space.transformation.map                             1600656.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . IsString.fromString        5167584.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . IsList.fromList            7683422.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . fromList                   7405281.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . unfoldr                     800596.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.fromList . intFromTo                7683422.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.write . intFromTo                   2065768.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . intFromTo                   800596.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.show                               35633488.00                      -0.00
All.Data.Array.Foreign/o-1-space.transformation.scanl'                          1638540.00                      -0.00
All.Data.Array.Foreign/o-1-space.transformation.scanl1'                         1638047.00                      -0.00
All.Data.Array.Foreign/o-n-space.elimination.writeLastN.Max                     1637294.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination.id                                  838187.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination.length . IsList.toList              838871.00                      -0.01
All.Data.Array.Foreign/o-1-space.elimination.min                                 838217.00                      -0.01
All.Data.Array.Foreign/o-1-space.elimination.<                                   838169.00                      -0.01
All.Data.Array.Foreign/o-n-space.elimination.writeN                              837156.00                      -0.01
All.Data.Array.Foreign/o-1-space.transformationX4.scanl1'                       4064377.00                      -1.58
All.Data.Array.Foreign/o-1-space.elimination.foldl'                              838233.00                      -2.45
All.Data.Array.Foreign/o-1-space.elimination.read                                838207.00                      -3.83

Data.Array.Foreign(bytesCopied)
Benchmark                                                                default(0)(Bytes) default(1) - default(0)(%)
------------------------------------------------------------------------ ----------------- --------------------------
All.Data.Array.Foreign/o-1-space.elimination.toStreamRev                          21858.00                      -0.18
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.1                             0.00                        NaN
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.10                            0.00                        NaN
All.Data.Array.Foreign/o-1-space.transformationX4.map                              1004.00                  +17162.05
All.Data.Array.Foreign/o-1-space.elimination./=                                   10961.00                     +98.99
All.Data.Array.Foreign/o-1-space.elimination.==                                   10962.00                     +87.97
All.Data.Array.Foreign/o-1-space.transformation.map                                 288.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . fromList                      4171.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . unfoldr                        144.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.fromList . intFromTo                   4618.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.write . intFromTo                       288.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . intFromTo                      144.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.show                               24596184.00                      -0.00
All.Data.Array.Foreign/o-1-space.generation.read                               20860416.00                      -0.01
All.Data.Array.Foreign/o-1-space.transformationX4.scanl'                         173801.00                      -0.14
All.Data.Array.Foreign/o-1-space.generation.writeN . IsString.fromString         153091.00                      -0.16
All.Data.Array.Foreign/o-1-space.transformation.scanl'                            21956.00                      -0.18
All.Data.Array.Foreign/o-1-space.elimination.length . IsList.toList               21938.00                      -0.18
All.Data.Array.Foreign/o-n-space.elimination.writeLastN.Max                       21897.00                      -0.18
All.Data.Array.Foreign/o-1-space.elimination.id                                   21853.00                      -0.18
All.Data.Array.Foreign/o-1-space.elimination.min                                  21835.00                      -0.18
All.Data.Array.Foreign/o-n-space.elimination.writeN                               21801.00                      -0.18
All.Data.Array.Foreign/o-1-space.transformation.scanl1'                           21894.00                      -0.19
All.Data.Array.Foreign/o-1-space.elimination.<                                    21833.00                      -0.19
All.Data.Array.Foreign/o-1-space.generation.writeN . IsList.fromList             153960.00                      -0.61
All.Data.Array.Foreign/o-1-space.elimination.foldl'                               21862.00                      -6.24
All.Data.Array.Foreign/o-1-space.elimination.read                                 21850.00                     -51.82
All.Data.Array.Foreign/o-1-space.transformationX4.scanl1'                          4025.00                    -321.47
```
Few measurement errors.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 3]
```diff
@@ -1452,22 +1439,37 @@ toListFB c n Array{..} = go arrStart
...
 --
 -- @toStreamD = D.unfold read@
 --
 -- We can try this if the unfold has any performance issues.
 {-# INLINE_NORMAL toStreamD #-}
-toStreamD :: forall m a. (Monad m, Storable a) => Array a -> D.Stream m a
+toStreamD :: forall m a. (MonadIO m, Storable a) => Array a -> D.Stream m a
 toStreamD Array{..} =
     let p = arrStart
     in D.Stream step p
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1475]: We can change it to `D.Stream step arrStart`.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 4]
```diff
@@ -1477,23 +1479,12 @@ toStreamD Array{..} =
...
-        -- evaluated/written to before we peek at them.
-        --
-        -- XXX This may not be safe for mutable arrays.
-        --
-        let !x = unsafeInlineIO $ do
-                    r <- peek p
-                    touch arrContents
-                    return r
-        return $ D.Yield x (p `plusPtr` sizeOf (undefined :: a))
+        !r <- liftIO $ peek p
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1482]: Bang can be dropped now.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 5]
```diff
@@ -1477,23 +1479,12 @@ toStreamD Array{..} =
...
+        liftIO $ touch arrContents
+        return $ D.Yield r (p `plusPtr` sizeOf (undefined :: a))
 
 {-# INLINE toStreamK #-}
-toStreamK :: forall m a. Storable a => Array a -> K.Stream m a
+toStreamK :: forall m a. (MonadIO m, Storable a) => Array a -> K.Stream m a
 toStreamK Array{..} =
     let p = arrStart
     in go p
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1490]: Can make it `go arrStart`.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 6]
```diff
@@ -1526,16 +1515,12 @@ toStreamDRev Array{..} =
...
     step _ p = do
-        -- See comments in toStreamD for why we use unsafeInlineIO
-        -- XXX
-        let !x = unsafeInlineIO $ do
-                    r <- peek p
-                    touch arrContents
-                    return r
-        return $ D.Yield x (p `plusPtr` negate (sizeOf (undefined :: a)))
+        r <- liftIO $ peek p
+        liftIO $ touch arrContents
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1519]: We can use a single liftIO call.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 7]
```diff
@@ -1477,23 +1479,12 @@ toStreamD Array{..} =
...
-        --
-        -- XXX This may not be safe for mutable arrays.
-        --
-        let !x = unsafeInlineIO $ do
-                    r <- peek p
-                    touch arrContents
-                    return r
-        return $ D.Yield x (p `plusPtr` sizeOf (undefined :: a))
+        !r <- liftIO $ peek p
+        liftIO $ touch arrContents
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1483]: We can use a single liftIO call.

File: src/Streamly/Internal/Data/Array/Foreign/Type.hs [hunk 8]
```diff
@@ -441,23 +442,103 @@ length arr = MA.length (unsafeThaw arr)
...
+                    touch arrContents
+                    return r
+        return $ D.Yield x (p `plusPtr` sizeOf (undefined :: a))
 
 {-# INLINE toStreamK #-}
 toStreamK :: forall m a. Storable a => Array a -> K.Stream m a
-toStreamK arr = MA.toStreamK (unsafeThaw arr)
+toStreamK Array{..} =
+    let p = arrStart
+    in go p
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Type.hs:496]: `go arrStart`

File: src/Streamly/Internal/Data/Array/Foreign/Type.hs [hunk 9]
```diff
@@ -441,23 +442,103 @@ length arr = MA.length (unsafeThaw arr)
...
+            let cur = p `plusPtr` negate (sizeOf (undefined :: a))
+            return $ D.Yield x (ReadUState contents start cur)
+
 
 {-# INLINE_NORMAL toStreamD #-}
 toStreamD :: forall m a. (Monad m, Storable a) => Array a -> D.Stream m a
-toStreamD arr = MA.toStreamD (unsafeThaw arr)
+toStreamD Array{..} =
+    let p = arrStart
+    in D.Stream step p
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Type.hs:472]: `D.Stream step arrStart`

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Comment by harendra-kumar** [comment]
`All.Data.Array.Foreign/o-1-space.transformationX4.map                              1004.00                  +17162.05`

Is this reproducible every time? This benchmarks seems to have regressed by 40% but allocations are the same. Also see if the regressions in writeLastN and `/=` are consistent or just test to test variations.

**Comment by adithyaov** [comment]
> All.Data.Array.Foreign/o-1-space.transformationX4.map 1004.00 +17162.05
>
> Is this reproducible every time? This benchmarks seems to have regressed by 40% but allocations are the same. Also see if the regressions in writeLastN and /= are consistent or just test to test variations.

Ok, I forgot to mention. The first one is this branch and compared with that is origin/master.
It is reproducible. I don't see why there's such an improvement though.

File: src/Streamly/Internal/Data/Array/Foreign.hs [hunk 10]
```diff
@@ -229,13 +249,11 @@ unsafeRead = Unfold step inject
             -- evaluated/written to before we peek at them.
             let !x = unsafeInlineIO $ do
                         r <- peek (Ptr p)
-                        touch contents
+                        touch (fptrToArrayContents contents)
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply src/Streamly/Internal/Data/Array/Foreign.hs:252]: I'm replacing  `ForeignPtr` with `ReadUState`

**Comment by adithyaov** [comment]
Updated bench report without `--quick`

```
Generating reports for Data.Array.Foreign...
Data.Array.Foreign(cpuTime)
Benchmark                                                                default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------ -------------- --------------------------
All.Data.Array.Foreign/o-1-space.generation.fromList . intFromTo                1719.68                      +7.50
All.Data.Array.Foreign/o-1-space.elimination.foldl'                              126.74                      +7.26
All.Data.Array.Foreign/o-1-space.elimination.==                                  113.35                      +4.31
All.Data.Array.Foreign/o-1-space.generation.writeN . IsString.fromString        1266.14                      +2.33
All.Data.Array.Foreign/o-1-space.transformationX4.scanl1'                        490.91                      +1.75
All.Data.Array.Foreign/o-1-space.generation.show                               39099.10                      +1.64
All.Data.Array.Foreign/o-1-space.elimination.id                                   72.11                      +0.87
All.Data.Array.Foreign/o-1-space.generation.read                              215447.00                      +0.82
All.Data.Array.Foreign/o-1-space.generation.writeN . fromList                   1523.54                      -1.19
All.Data.Array.Foreign/o-1-space.elimination.read                                111.73                      -1.51
All.Data.Array.Foreign/o-1-space.transformationX4.map                            520.90                      -1.75
All.Data.Array.Foreign/o-1-space.generation.writeN . IsList.fromList            1798.75                      -2.07
All.Data.Array.Foreign/o-1-space.elimination.<                                   175.45                      -2.57
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.1                        115.29                      -2.61
All.Data.Array.Foreign/o-n-space.elimination.writeLastN.Max                      219.67                      -2.68
All.Data.Array.Foreign/o-1-space.elimination.length . IsList.toList              195.63                      -3.33
All.Data.Array.Foreign/o-1-space.generation.writeN . intFromTo                    73.65                      -3.61
All.Data.Array.Foreign/o-1-space.elimination.min                                 178.18                      -4.12
All.Data.Array.Foreign/o-1-space.transformation.map                              177.16                      -4.29
All.Data.Array.Foreign/o-1-space.generation.writeN . unfoldr                     110.76                      -4.76
All.Data.Array.Foreign/o-n-space.elimination.writeN                              108.77                      -5.39
All.Data.Array.Foreign/o-1-space.transformation.scanl1'                          182.74                      -5.48
All.Data.Array.Foreign/o-1-space.transformation.scanl'                           188.99                      -6.03
All.Data.Array.Foreign/o-1-space.transformationX4.scanl'                         519.48                      -7.29
All.Data.Array.Foreign/o-1-space.elimination./=                                  116.37                      -8.66
All.Data.Array.Foreign/o-1-space.generation.write . intFromTo                    204.76                      -9.99
All.Data.Array.Foreign/o-1-space.elimination.toStreamRev                         122.04                     -12.61
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.10                       150.48                     -33.01

Data.Array.Foreign(Allocated)
Benchmark                                                                default(0)(Bytes) default(1) - default(0)(%)
------------------------------------------------------------------------ ----------------- --------------------------
All.Data.Array.Foreign/o-1-space.generation.fromList . intFromTo                8809784.00                      +0.10
All.Data.Array.Foreign/o-1-space.generation.writeN . fromList                   7993141.00                      +0.04
All.Data.Array.Foreign/o-1-space.elimination.<                                   800594.00                      +0.00
All.Data.Array.Foreign/o-1-space.elimination.==                                  800594.00                      +0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . unfoldr                     800553.00                      +0.00
All.Data.Array.Foreign/o-1-space.elimination.toStreamRev                         800552.00                      +0.00
All.Data.Array.Foreign/o-1-space.generation.show                               35523344.00                      +0.00
All.Data.Array.Foreign/o-1-space.elimination.min                                 800632.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.write . intFromTo                   2065613.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . intFromTo                   800553.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.read                              542617654.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination.foldl'                              800554.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination.id                                  800554.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination.length . IsList.toList              800616.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination./=                                  800600.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination.read                                800560.00                      -0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . IsString.fromString        5619260.00                      -0.06
All.Data.Array.Foreign/o-1-space.generation.writeN . IsList.fromList            8794774.00                      -0.25
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.1                             0.00                        NaN
All.Data.Array.Foreign/o-n-space.elimination.writeN                              800556.00                       0.00
All.Data.Array.Foreign/o-1-space.transformation.map                             1600576.00                       0.00
All.Data.Array.Foreign/o-1-space.transformation.scanl1'                         1600570.00                       0.00
All.Data.Array.Foreign/o-1-space.transformationX4.map                           4000621.00                      -0.00
All.Data.Array.Foreign/o-n-space.elimination.writeLastN.Max                     1600608.00                      -0.00
All.Data.Array.Foreign/o-1-space.transformationX4.scanl1'                       4000639.00                      -0.00
All.Data.Array.Foreign/o-1-space.transformationX4.scanl'                        4000639.00                      -0.00
All.Data.Array.Foreign/o-1-space.transformation.scanl'                          1600576.00                      -0.00
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.10                          405.00                  -Infinity

Data.Array.Foreign(bytesCopied)
Benchmark                                                                default(0)(Bytes) default(1) - default(0)(%)
------------------------------------------------------------------------ ----------------- --------------------------
All.Data.Array.Foreign/o-1-space.elimination.<                                       74.00                      +4.05
All.Data.Array.Foreign/o-1-space.elimination.==                                      74.00                      +4.05
All.Data.Array.Foreign/o-1-space.elimination.toStreamRev                             72.00                      +2.78
All.Data.Array.Foreign/o-1-space.generation.writeN . unfoldr                         73.00                      +1.37
All.Data.Array.Foreign/o-1-space.generation.fromList . intFromTo                 369115.00                      +0.64
All.Data.Array.Foreign/o-1-space.generation.writeN . fromList                    187093.00                      +0.15
All.Data.Array.Foreign/o-1-space.generation.show                               22033360.00                      +0.00
All.Data.Array.Foreign/o-1-space.elimination.min                                     77.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.write . intFromTo                       149.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.writeN . intFromTo                       74.00                       0.00
All.Data.Array.Foreign/o-1-space.generation.read                               20860367.00                      -0.01
All.Data.Array.Foreign/o-1-space.generation.writeN . IsString.fromString         293868.00                      -0.62
All.Data.Array.Foreign/o-1-space.elimination.foldl'                                  74.00                      -1.37
All.Data.Array.Foreign/o-1-space.elimination.id                                      74.00                      -1.37
All.Data.Array.Foreign/o-1-space.generation.writeN . IsList.fromList             365952.00                      -1.77
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.1                             0.00                        NaN
All.Data.Array.Foreign/o-1-space.elimination.writeLastN.10                            0.00                        NaN
All.Data.Array.Foreign/o-n-space.elimination.writeLastN.Max                         149.00                      +3.36
All.Data.Array.Foreign/o-n-space.elimination.writeN                                  74.00                       0.00
All.Data.Array.Foreign/o-1-space.transformation.map                                 149.00                       0.00
All.Data.Array.Foreign/o-1-space.transformation.scanl1'                             149.00                       0.00
All.Data.Array.Foreign/o-1-space.transformationX4.map                               362.00                      -1.40
All.Data.Array.Foreign/o-1-space.transformation.scanl'                              149.00                      -2.76
All.Data.Array.Foreign/o-1-space.transformationX4.scanl1'                           373.00                      -3.04
All.Data.Array.Foreign/o-1-space.transformationX4.scanl'                            373.00                      -3.04
All.Data.Array.Foreign/o-1-space.elimination.read                                    77.00                      -4.05
All.Data.Array.Foreign/o-1-space.elimination.length . IsList.toList                  77.00                      -4.05
All.Data.Array.Foreign/o-1-space.elimination./=                                      77.00                      -4.05
```

**Comment by harendra-kumar** [comment]
Till now mutable array was mostly getting benchmarked via immutable array, but now it has diverged in most operations, so I think we need to have a separate benchmark suite for mutable arrays. Can you create an issue for that?

**Review by harendra-kumar — APPROVED** [review]

**Pull request merged**